src/core/core/model/story.py

```diff
@@ -379,13 +379,75 @@ def get_for_worker(cls, filter_args: dict) -> list[dict[str, Any]]:
         return stories
 
     @classmethod
-    def add(cls, data) -> "Story":
-        item = cls.from_dict(data)
-        db.session.add(item)
-        db.session.commit()
-        StorySearchIndex.prepare(item)
-        item.update_tlp()
-        return item
+    def add(cls, data) -> "tuple[Story | None, dict, int]":
+        def filter_news_items(story):
+            original_count = len(story.news_items)
+            story.news_items = [ni for ni in story.news_items if not NewsItem.identical(ni.hash)]
+            skip_count = original_count - len(story.news_items)
+            if skip_count > 0:
+                logger.info(f"Skipped {skip_count} identical news items.")
+            return original_count, skip_count, bool(story.news_items)
+
+        try:
+            story = cls.from_dict(data)
+            original_count, skip_count, has_new_items = filter_news_items(story)
+
+            if not has_new_items:
+                logger.info("No items to add to the story. Skipping database insert.")
+                return None, {"error": "All items are duplicates"}, 409
+
+            if skip_count > 0 and skip_count < original_count:
+                logger.info("Partial match detected. Update the story instead.")
+                return None, {"error": "Partial match detected. Update the story instead."}, 422
+
+            db.session.add(story)
+            db.session.commit()
+
+            StorySearchIndex.prepare(story)
+            story.update_tlp()
+
+            logger.info(f"Story added successfully: {story.id}")
+            return story, {"message": "Story added successfully"}, 200
+
+        except Exception as e:
+            logger.exception(f"Failed to process story: {e}")
+            db.session.rollback()
+            return None, {"error": "An unexpected error occurred"}, 500
+
+    @classmethod
+    def update_story_cluster(cls, data) -> tuple[dict, int]:
+        try:
+            existing_story_id = None
+            missing_news_items = []
+            for news_item in data.get("news_items", []):
+                if item := NewsItem.find_by_hash(news_item.get("hash")):
+                    existing_story_id = item[0].story_id
+                else:
+                    missing_news_items.append(news_item)
```

Comment:
Story id should never be extracted like that.

It has to be provided as input `data['id']`